design/src/docs/anatomy.md

```diff
@@ -0,0 +1,687 @@
+        let layer = Stack::new(UpgradeLayer::new(), layer);
+        layer.layer(inner)
+    }
+}
+```
+
+Why do we need a trait for this? Why not simply write an `upgrade` method on `Operation<S, L>`? The reason is that we might _not_ want to supply an `Operation<S, L>` to the service builder, instead we might want to supply something that overrides the typical upgrade procedure.
+
+Below we give an example of a ZST which can be provided to the builder, which also satisfies `Upgradable` and returns a `MissingFailure` `tower::Service`. This `MissingFailure` service simply returns a status code 500.
+
+```rust
+/// A marker struct indicating an [`Operation`] has not been set in a builder.
+///
+/// This _does_ implement [`Upgradable`] but produces a [`Service`] which always returns an internal failure message.
+pub struct FailOnMissingOperation;
+
+impl<P, Op, B> Upgradable<P, Op> for FailOnMissingOperation
+where
+    InternalFailureException: IntoResponse<P>,
+{
+    type Service = MissingFailure<P>;
+
+    fn upgrade(self, _plugin: &Pl) -> Self::Service {
+        MissingFailure { _protocol: PhantomData }
+    }
+}
+```
+
+We go into more detail on how the `Upgradable` trait is used in conjunction with builders in the [Builders](#builders) section below.
+
+## Routers
+
+Different protocols supported by Smithy enjoy different routing mechanisms, for example, [AWS JSON 1.0](https://awslabs.github.io/smithy/2.0/aws/protocols/aws-json-1_0-protocol.html#protocol-behaviors) uses the `X-Amz-Target` header to select an operation, whereas [AWS REST XML](https://awslabs.github.io/smithy/2.0/aws/protocols/aws-restxml-protocol.html) uses the [HTTP label trait](https://awslabs.github.io/smithy/2.0/spec/http-bindings.html#httplabel-trait).
+
+Despite their differences, all routing mechanisms satisfy a common interface. This is formalized using the `Router` trait:
+
+```rust
+/// An interface for retrieving an inner [`Service`] given a [`http::Request`].
+pub trait Router {
+    type Service;
+    type Error;
+
+    /// Matches a [`http::Request`] to a target [`Service`].
+    fn match_route(&self, request: &http::Request) -> Result<Self::Service, Self::Error>;
+}
+```
+
+which provides the ability to lookup an inner HTTP service from a collection using a `&http::Request`.
+
+Types which implement the `Router` trait are converted to a HTTP service via the `RoutingService` struct:
+
+```rust
+/// A [`Service`] using a [`Router`] `R` to redirect messages to specific routes.
+///
+/// The `Protocol` parameter is used to determine the serialization of errors.
+pub struct RoutingService<R, Protocol> {
+    router: R,
+    _protocol: PhantomData<Protocol>,
+}
+
+impl<R, P> Service<http::Request> for RoutingService<R, P>
+where
+    R: Router<B>,
+    R::Service: Service<http::Request, Response = http::Response>,
+    R::Error: IntoResponse<P> + Error,
+{
+    type Response = http::Response;
+    type Error = Box<dyn Error>;
```

Comment:
I'd call out again that we're simplifying the implementation in favor of a more pedagogical text, lest a reader think we're actually boxing every time we return a routing error.